lte/cloud/go/services/subscriberdb/storage/ip_lookup.go

```diff
@@ -0,0 +1,143 @@
+/*
+ Copyright 2020 The Magma Authors.
+
+ This source code is licensed under the BSD-style license found in the
+ LICENSE file in the root directory of this source tree.
+
+ Unless required by applicable law or agreed to in writing, software
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+*/
+
+package storage
+
+import (
+	"database/sql"
+	"fmt"
+	"sort"
+
+	"magma/lte/cloud/go/services/subscriberdb/protos"
+	"magma/lte/cloud/go/services/subscriberdb/state"
+	"magma/orc8r/cloud/go/sqorc"
+
+	"github.com/Masterminds/squirrel"
+	"github.com/pkg/errors"
+)
+
+type IPLookup interface {
+	// Initialize the backing store.
+	Initialize() error
+
+	// GetIPs returns the set of IMSIs each IP is marked as assigned to, keyed
+	// by IP address.
+	GetIPs(networkID string, ips []string) ([]*protos.IPMapping, error)
+
+	// SetIPs assigns an IP to an IMSI under a particular APN, one per mapping.
+	SetIPs(networkID string, mappings []*protos.IPMapping) error
+}
+
+const (
+	tableName = "subscriberdb_ip_to_imsi"
+
+	nidCol  = "network_id"
+	ipCol   = "ip"
+	imsiCol = "imsi_and_apn"
+)
+
+type ipLookup struct {
+	db      *sql.DB
+	builder sqorc.StatementBuilder
+}
+
+func NewIPLookup(db *sql.DB, builder sqorc.StatementBuilder) IPLookup {
+	return &ipLookup{db: db, builder: builder}
+}
+
+func (l *ipLookup) Initialize() error {
+	txFn := func(tx *sql.Tx) (interface{}, error) {
+		_, err := l.builder.CreateTable(tableName).
+			IfNotExists().
+			Column(nidCol).Type(sqorc.ColumnTypeText).NotNull().EndColumn().
+			Column(ipCol).Type(sqorc.ColumnTypeText).NotNull().EndColumn().
+			Column(imsiCol).Type(sqorc.ColumnTypeText).NotNull().EndColumn().
+			PrimaryKey(ipCol, imsiCol).
```

Comment:
I think maybe you should PK on the 3-tuple of (nid, ip, imsi.apn). all your queries start with nid so it'll make this a little more efficient. don't forget to update the upsert on conflict clause.

or not, we can bootcamp this later if we ever get to a scale where this matters